protocol/nats_jetstream/v3/receiver.go

```diff
@@ -0,0 +1,229 @@
+/*
+ Copyright 2024 The CloudEvents Authors
+ SPDX-License-Identifier: Apache-2.0
+*/
+
+package nats_jetstream
+
+import (
+	"context"
+	"io"
+	"sync"
+
+	"github.com/nats-io/nats.go"
+	"github.com/nats-io/nats.go/jetstream"
+
+	"github.com/cloudevents/sdk-go/v2/binding"
+	"github.com/cloudevents/sdk-go/v2/protocol"
+)
+
+type msgErr struct {
+	msg binding.Message
+}
+
+type receiver struct {
+	incoming      chan msgErr
+	subMtx        sync.Mutex
+	internalClose chan struct{}
+
+	// nats connection
+	conn      *nats.Conn
+	url       string
+	natsOpts  []nats.Option
+	connOwned bool
+
+	// jetstream options and configuration
+	jetSteamOpts          []jetstream.JetStreamOpt
+	jetStream             jetstream.JetStream
+	consumerConfig        *jetstream.ConsumerConfig
+	orderedConsumerConfig *jetstream.OrderedConsumerConfig
+	pullConsumeOpts       []jetstream.PullConsumeOpt
+	jetstreamConsumer     jetstream.Consumer
+}
+
+// NewReceiver consumes from filterSubjects given in ConsumerConfig or OrderedConsumerConfig
+// embedded in ConsumerOption.
+// Use WithURL() OR WithConnection()
+// Use WithConsumerConfig() OR WithOrderedConsumerConfig()
+func NewReceiver(ctx context.Context, opts ...ProtocolOption) (*receiver, error) {
+	c := &receiver{
+		incoming:      make(chan msgErr),
+		internalClose: make(chan struct{}, 1),
+	}
+
+	if err := c.applyOptions(opts...); err != nil {
+		return nil, err
+	}
+
+	if err := c.validateOptions(); err != nil {
+		return nil, err
+	}
+
+	var errConnection error
+	defer func() {
+		// close connection if an error occurs and we created the nats connection
+		if c.connOwned && c.conn != nil && errConnection != nil {
+			c.conn.Close()
+		}
+	}()
+
+	// create a connection based on the URL if the connection was not given as an option
+	if c.conn == nil {
+		c.conn, errConnection = nats.Connect(c.url, c.natsOpts...)
+		if errConnection != nil {
+			return nil, errConnection
+		}
+	}
+
+	if c.jetStream, errConnection = jetstream.New(c.conn, c.jetSteamOpts...); errConnection != nil {
+		return nil, errConnection
+	}
+
+	return c, nil
+}
+
+// MsgHandler implements nats.MsgHandler and publishes messages onto our internal incoming channel to be delivered
+// via r.Receive(ctx)
+func (r *receiver) MsgHandler(msg jetstream.Msg) {
+	r.incoming <- msgErr{msg: NewMessage(msg)}
+}
+
+// Receive implements Receiver.Receive.
+func (r *receiver) Receive(ctx context.Context) (binding.Message, error) {
+	select {
+	case msgErr, ok := <-r.incoming:
+		if !ok {
+			return nil, io.EOF
+		}
+		return msgErr.msg, nil
+	case <-ctx.Done():
+		return nil, io.EOF
+	}
+}
+
+// OpenInbound implements Opener.OpenInbound
+func (r *receiver) OpenInbound(ctx context.Context) error {
+	r.subMtx.Lock()
```

Comment:
Since only one protocol will ever use sender/receiver, you can implement locking at the protocol level instead of lower fields/methods here. WDYT? Makes locking simpler, e.g., during Close as well

Reply:
Please see other protocol implementations e.g., HTTP/Kafka (Confluent) which use a more simpler flow to reduce protocol complexity